Grapejuice v7.14.4 & Replace old Wine build

guihkx opened the pull request:
This updates Grapejuice to [v7.14.4](https://gitlab.com/brinkervii/grapejuice/-/releases/v7.14.4) and switches from the old, git-based data checker to a JSON-based one using GitLab's API, which should be more reliable for detecting future releases.

This also replaces the old Wine-Tkg build (v7.20) by a 'vanilla', Wine Staging build (v8.13) [provided by Kron4ek](https://github.com/Kron4ek/Wine-Builds).

Keep in mind that even after this upgrade, I still can't even get past the installer screen of Roblox/Studio on a clean prefix. Terminating the MSEdge process also terminates the installer with an error ("Wine is not supported").

Regardless of that, I still think this is an upgrade worth doing...

Before merging this though, it'd be good to get at least one confirmation that this major Wine upgrade won't break people's (working?) prefixes.

Closes #106
Closes #107 
Closes #108 
Closes #109

guihkx committed:
Update Grapejuice to v7.14.4

Also switch to a JSON-based API for Flathub's external data checker,
since the previous one (git-based) wasn't working, for some reason.

Release notes for v7.14.4:

This version of Grapejuice adds a Roblox Release channel setting!

Aside that some minor bugs were fixed:
- Grapejuice now throws a real error when it can not fetch a version
- The self-updater doesn't crash anymore. This broke after the virtualenv update.
- Removed the player warning dialog

guihkx committed:
Update Grapejuice dependencies

This also pins pydantic to version '<2', which has less dependencies.

guihkx committed:
Update build options for glxinfo

Just to speed up compilation a bit. And also remove a leftover
directory in /share.

guihkx commented:
This surprises me. I switched to the 'Integration' channel and I was suddenly able to play the game again. I don't know for how long, though. :)

![image](https://github.com/flathub/net.brinkervii.grapejuice/assets/626206/bc1700e3-4c78-461c-b412-5045b76b573d)

TheComputerGuy96 reviewed (approved):
Everything seems to be fine (I'll probably push a better Wine build though) 🐸

guihkx commented:
Thanks for testing!

> (I'll probably push a better Wine build though) 🐸

Nice. Assuming that build already exists, feel free to let me know where I can get it, and I'll update this PR. :+1:

TheNexusAvenger commented:
> To test this build, install it from the testing repository:
> 
> ```
> flatpak install --user https://dl.flathub.org/build-repo/42303/net.brinkervii.grapejuice.flatpakref
> ```

Any idea if this command is meant to work as-is? I tried it on SteamOS, and I get an error message saying that WINE doesn't exist when running `subprocess`. That file is definitely there in Flatpak, otherwise, a different error message saying WINE was not found would have happened first.

![image](https://github.com/flathub/net.brinkervii.grapejuice/assets/13441476/2ef5d32a-b569-46ef-81f0-c64ce615143b)

guihkx commented:
>Any idea if this command is meant to work as-is?

It is. I'm thinking your issue is related to a modified `PATH` variable (something I actually experienced myself).

Can you check if you have any Flatpak overrides for Grapejuice? I.e.:

```
cat ~/.local/share/flatpak/overrides/net.brinkervii.grapejuice
```

If the file doesn't exist, then the problem is something else.

If it does exist, then I think I'll have to install Wine to `/app/patched_wine` instead of switching to `/app/wine-staging`, just to avoid breaking people's setup.

TheNexusAvenger commented:
> Can you check if you have any Flatpak overrides for Grapejuice? I.e.:
> 
> ```
> cat ~/.local/share/flatpak/overrides/net.brinkervii.grapejuice
> ```

On my Steam Deck, I don't even have an overrides directory. The WINE files are there in the package; [`CouldNotFindSystemWineHome` would have been thrown first before this error otherwise](https://gitlab.com/brinkervii/grapejuice/-/blob/master/src/grapejuice_common/variables.py#L84). When I use flathubbot install command or a build the master branch normally on another system using [their first build instructions](https://docs.flatpak.org/en/latest/first-build.html), I get this issue with `subprocess.call` saying WINE doesn't exist.

guihkx commented:
I've never experienced/seen something like that, so I have no clue what it could be... :man_shrugging: 

Can you post the entire Grapejuice traceback as text here?

Also, can you double check if `/app/bin/wine-staging/bin` really exists within the Flatpak sandbox (you never know, lol)? Run these commands and see if their output matches mine:

```
$ flatpak run --command=which net.brinkervii.grapejuice//test wine
/app/wine-staging/bin/wine
$ flatpak run --command=stat net.brinkervii.grapejuice//test /app/wine-staging/bin/wine
  File: /app/wine-staging/bin/wine
  Size: 9540            Blocks: 24         IO Block: 4096   regular file
Device: 254,0   Inode: 1358024     Links: 3
Access: (0755/-rwxr-xr-x)  Uid: (65534/nfsnobody)   Gid: (65534/nfsnobody)
Access: 2023-08-19 09:36:10.960476588 -0300
Modify: 1969-12-31 21:00:00.000000000 -0300
Change: 2023-08-19 16:39:53.444222850 -0300
 Birth: 2023-08-19 09:35:53.007442159 -0300
$ flatpak run --command=wine net.brinkervii.grapejuice//test --version
wine-8.13 (Staging)
```

TheNexusAvenger commented:
> Can you post the entire Grapejuice traceback as text here?

As requested, although the next request has more interesting results:
```
Traceback (most recent call last):
  File "/app/lib/python3/dist-packages/grapejuice/background.py", line 82, in run
    self.work()
  File "/app/lib/python3/dist-packages/grapejuice/tasks.py", line 43, in work
    abstract_grapejuice().launch_app(self._prefix.configuration.id)
  File "/app/lib/python3/dist-packages/grapejuice_common/abstraction/abstract_grapejuice_impl.py", line 32, in launch_app
    _with_prefix_id(prefix_id, lambda prefix: prefix.roblox.launch_app())
  File "/app/lib/python3/dist-packages/grapejuice_common/abstraction/abstract_grapejuice_impl.py", line 19, in _with_prefix_id
    prefix.roblox.install_roblox(post_install_function=lambda: cb(prefix))
  File "/app/lib/python3/dist-packages/grapejuice_common/wine/wineprefix_roblox.py", line 64, in install_roblox
    self._core_control.create_prefix()
  File "/app/lib/python3/dist-packages/grapejuice_common/wine/wineprefix_core_control.py", line 436, in create_prefix
    self.configure_prefix()
  File "/app/lib/python3/dist-packages/grapejuice_common/wine/wineprefix_core_control.py", line 432, in configure_prefix
    self.disable_mime_associations()
  File "/app/lib/python3/dist-packages/grapejuice_common/wine/wineprefix_core_control.py", line 419, in disable_mime_associations
    self.load_registry_file(paths.assets_directory() / "disable_mime_assoc.reg")
  File "/app/lib/python3/dist-packages/grapejuice_common/wine/wineprefix_core_control.py", line 391, in load_registry_file
    self.run_exe("regedit", "/S", winreg, run_async=False, use_wine64=False)
  File "/app/lib/python3/dist-packages/grapejuice_common/wine/wineprefix_core_control.py", line 471, in run_exe
    return do_run_exe(
  File "/app/lib/python3/dist-packages/grapejuice_common/logs/log_util.py", line 13, in wrapper
    result = func(*args, **kwargs)
  File "/app/lib/python3/dist-packages/grapejuice_common/wine/wineprefix_core_control.py", line 158, in do_run_exe
    subprocess.call(
  File "/usr/lib/python3.10/subprocess.py", line 345, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/usr/lib/python3.10/subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.10/subprocess.py", line 1863, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/app/wine-staging/bin/wine'
```

> Also, can you double check if `/app/bin/wine-staging/bin` really exists within the Flatpak sandbox (you never know, lol)? Run these commands and see if their output matches mine:

This may be more interesting. The last command also gives "No such file or directory".

```
(deck@steamdeck ~)$ flatpak run --command=which net.brinkervii.grapejuice//test wine
/app/wine-staging/bin/wine
(deck@steamdeck ~)$ flatpak run --command=stat net.brinkervii.grapejuice//test /app/wine-staging/bin/wine
  File: /app/wine-staging/bin/wine
  Size: 9540            Blocks: 24         IO Block: 4096   regular file
Device: 259,8   Inode: 4296269     Links: 2
Access: (0755/-rwxr-xr-x)  Uid: ( 1000/    deck)   Gid: ( 1000/    deck)
Access: 2023-08-19 15:35:35.163439390 -0400
Modify: 1969-12-31 19:00:00.000000000 -0500
Change: 2023-08-19 15:36:41.908810842 -0400
 Birth: 2023-08-19 15:35:35.163439390 -0400
(deck@steamdeck ~)$ flatpak run --command=wine net.brinkervii.grapejuice//test --version
bwrap: execvp wine: No such file or directory
```

Edit: In addition
```
(deck@steamdeck ~)$ flatpak run --command=/app/wine-staging/bin/wine net.brinkervii.grapejuice//test --version
bwrap: execvp /app/wine-staging/bin/wine: No such file or directory
```

guihkx commented:
What the hell. :smiling_face_with_tear: 

I remember experiencing something like that before, and if I recall correctly, it was due to the binary being compiled with weird/incompatible flags for my CPU.

But given that Steam Deck's CPU is *much* newer than mine (an Intel i5 3570k), I think that's not it.

Just for fun, try running this:

```
flatpak run --command=wine --env=LD_DEBUG=all net.brinkervii.grapejuice//test --version
```

And post the output here (as an attachment if it's too large, please).

TheNexusAvenger commented:
> I remember experiencing something like that before, and if I recall correctly, it was due to the binary being compiled with weird/incompatible flags for my CPU.
> 
> But given that Steam Deck's CPU is much newer than mine (an Intel i5 3570k), I think that's not it.

Interesting. Both my personal system and the Steam Deck are based on AMD Zen 2.
I'll set up a quick Manjaro VM on my desktop and see what happens with the current Flatpak and the new one.

> Just for fun, try running this:
> 
> ```
> flatpak run --command=wine --env=LD_DEBUG=all net.brinkervii.grapejuice//test --version
> ```

Sadly, same output.
```
(deck@steamdeck ~)$ flatpak run --command=wine --env=LD_DEBUG=all net.brinkervii.grapejuice//test --version
bwrap: execvp wine: No such file or directory
```

lunarlattice0 commented:
@guihkx I am waiting on Brinker/@TheComputerGuy96 to package wine-tkg 8.X; we have discovered in the past that vanilla wine does not play well with Grapejuice.

TheNexusAvenger commented:
> I'll set up a quick Manjaro VM on my desktop and see what happens with the current Flatpak and the new one.

Whatever is going on seems to be SteamOS-specific. WINE is able to start Roblox on a clean install of Manjaro using both current and new versions of the Flatpak. I'll definitely need to re-test with wine-tkg once that is done.

guihkx commented:
> Sadly, same output.

I just learned that `/app/wine-staging/bin/wine` is a 32-bit binary:

```
$ flatpak run --command=file net.brinkervii.grapejuice//test /app/wine-staging/bin/wine
/app/wine-staging/bin/wine: ELF 32-bit LSB pie executable, Intel 80386, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux.so.2, for GNU/Linux 3.2.0, BuildID[sha1]=c7927e8dd71b4b2573986082ad1c7d7806c7b6e5, stripped
```

Which is not a big deal, since we have correctly enabled [`multiarch` support](https://docs.flatpak.org/en/latest/multiarch.html).

However, it's possible that the Flatpak version present on the Deck is a bit old and, for some reason, doesn't mount the required directory where 32-bit libraries reside, when the target branch of the app is anything other than `stable` (in this case it's called `test`).

For reference, on Arch the current version of Flatpak is `1.15.4`.

And to confirm this theory, you could try making a local build that has `branch: stable` added at the top of `net.brinkervii.grapejuice.yml`. If my theory is correct, then things should work.

But then again, I could be wrong and it might be something else - for example, the fact that on the Deck you kinda have to install everything user-wide (due to its immutable filesystem)  instead of system-wide, and that somehow causes issues.

**EDIT:** You can also check if `org.freedesktop.Platform.Compat.i386` is actually installed:

```
$ flatpak info org.freedesktop.Platform.Compat.i386
          ID: org.freedesktop.Platform.Compat.i386
         Ref: runtime/org.freedesktop.Platform.Compat.i386/x86_64/22.08
        Arch: x86_64
      Branch: 22.08
      Origin: flathub
  Collection: org.flathub.Stable
Installation: system
   Installed: 258,6 MB
```

Which should be done automatically when you install the package.

---

@Thelolguy1 Alright, no problem. I'm not sure how you want to proceed now, lol. Should I abandon this pull request altogether? Or should I wait until you folks make a build for me to incorporate here?

TheNexusAvenger commented:
> **EDIT:** You can also check if `org.freedesktop.Platform.Compat.i386` is actually installed:
> 
> ```
> $ flatpak info org.freedesktop.Platform.Compat.i386
>           ID: org.freedesktop.Platform.Compat.i386
>          Ref: runtime/org.freedesktop.Platform.Compat.i386/x86_64/22.08
>         Arch: x86_64
>       Branch: 22.08
>       Origin: flathub
>   Collection: org.flathub.Stable
> Installation: system
>    Installed: 258,6 MB
> ```
> 
> Which should be done automatically when you install the package.

My Steam Deck had 2**1**.08 instead of 22.08. Manually adding 22.08 seems to have resolved that issue. Needing to kill the WebView2 process, [as covered in this issue](https://gitlab.com/brinkervii/grapejuice/-/issues/492), to get the bootstrapper to continue was also needed.

guihkx committed:
Minor tweaks to multiarch support

Just to sync with the official documentation:

https://docs.flatpak.org/en/latest/multiarch.html

- Add the debug extension for i386
- Remove ld.so.conf and write one directly in the manifest

guihkx commented:
>Manually adding 22.08 seems to have resolved that issue

It's weird that you had to manually install it, but I'm glad you got it working. :)

>Needing to kill the WebView2 process, [as covered in this issue](https://gitlab.com/brinkervii/grapejuice/-/issues/492)

Yeah, I had to do the same. By the way, I can't launch the game anymore, for some reason...

In any case, I think the main objective of this pull request is to provide a recent build of Wine (while hopefully not breaking people's prefixes), and not really getting Roblox back to a working state.

guihkx commented:
Hmm, I'm missing app icons here (even on the `stable` build, actually):

![image](https://github.com/flathub/net.brinkervii.grapejuice/assets/626206/4adcc111-3c25-4f13-8c47-213f5000f216)

I'll investigate...

guihkx committed:
Copy and rename icons instead of symlinking them

Symbolic links pointing to the original icons didn't work as intended,
because the original icons wouldn't get exported by flatpak-builder due
to their 'non-allowed' naming scheme.

Instead, we should just make copies of these icons and rename them
appropriately for Flatpak.

Even though the original icons won't be exported by flatpak-builder, we
should still keep them around because Grapejuice expects them there.

guihkx committed:
Switch to a Wine Staging build

We were using an ancient build of Wine-Tkg, which is now unrecommended
by upstream:

https://gitlab.com/brinkervii/grapejuice/-/commit/af62d6ef47dab4faa6407b6e11bab75c80e1bb3a

The switch to the staging build also gets rid of a new error given by
the installer of both Roblox and Roblox Studio:

"Roblox does not support Admin installation, please rerun
RobloxPlayerInstaller without admin privilege."

The Wine Staging build is provided by Kron4ek:

https://github.com/Kron4ek/Wine-Builds

lunarlattice0 commented:
@guihkx Don't abandon this PR, you've made a lot of good progress. I will let you know when Brinker has a wine-tkg build available.

guihkx commented:
@Thelolguy1 Perfect! I'll be waiting then.

---

Icons have been fixed:

![image](https://github.com/flathub/net.brinkervii.grapejuice/assets/626206/0fc43aee-94ed-4e95-8459-3ac91db5e930)

TheNexusAvenger commented:
Great to see the progress being made. SteamOS is working, although not issue-free, because Roblox.
I found out that an older `org.freedesktop.Platform.Compat.i386` is from another Flatpak I didn't need anymore. Removing it so no version of `org.freedesktop.Platform.Compat.i386` was installed changed nothing; clean installing the Flatpak didn't make it automatically install.

By any chance, does adding something like the following do anything useful?
```
sdk-extensions:
  - org.freedesktop.Platform.Compat.i386
```

I don't know enough about Flatpak to understand what this does. On a local build, it prevented the build from going without me installing it. That would prevent me from going through the pain of the error message mentioned in my first comment.

**Edit**: It does **not** function as I was hoping it would. There is no way to satisfy the dependency.

**Edit 2**: Including `org.freedesktop.Sdk.Compat.i386` did nothing useful, but I may be testing it incorrectly.

guihkx commented:
>clean installing the Flatpak didn't make it automatically install.

Unfortunately, it looks like that's a known, long-standing issue:

- https://github.com/flatpak/flatpak/issues/3072
- https://github.com/flatpak/flatpak/issues/3094
- https://github.com/flathub/net.lutris.Lutris/issues/53

TheNexusAvenger commented:
Lutris somehow handles this, although with `org.gnome.Platform.Compat.i386` instead of `org.freedesktop.Platform.Compat.i386`. There's got to be a way to make this work. Even if it isn't this PR, this will need to be handled *eventually* once kids with new Steam Decks become the target market.

![VirtualBoxVM_LvhMuNoGed](https://github.com/flathub/net.brinkervii.grapejuice/assets/13441476/68f3d4fb-0371-4267-b25d-64779ff8c97b)

guihkx commented:
What if you run `flatpak install flathub net.brinkervii.grapejuice`, wouldn't `org.freedesktop.Platform.Compat.i386` get pulled in as well?

TheNexusAvenger commented:
> What if you run `flatpak install flathub net.brinkervii.grapejuice`, wouldn't `org.freedesktop.Platform.Compat.i386` get pulled in as well?

It... does, which is good... but is the flathubbot install not including it...?

![VirtualBoxVM_kskCxLc1B6](https://github.com/flathub/net.brinkervii.grapejuice/assets/13441476/77a6c8f6-797a-4f51-bad1-71d12bb1b625)

guihkx commented:
>but is the flathubbot install not including it...?

I'm not sure, but maybe it's because builds from pull requests are put on the `test` branch, instead of `stable`.

lunarlattice0 commented:
@guihkx @TheNexusAvenger You can temporarily fix the missing dependencies issue by installing another Flatpak which uses the same dependency (e.g. org.winehq.Wine). This is a known bug unfortunately...

lunarlattice0 commented:
@guihkx OK, we've agreed to just use standard wine at the moment. Are you ready for me to accept this PR?

guihkx commented:
> @guihkx OK, we've agreed to just use standard wine at the moment. Are you ready for me to accept this PR?

Everything's good on my side, I think.

By the way, I opted to use Wine Staging but Kron4ek's repo also offers Wine-Tkg, so let me know if you prefer that instead. Otherwise, feel free to merge.

TheNexusAvenger commented:
> I'm not sure, but maybe it's because builds from pull requests are put on the `test` branch, instead of `stable`.

> @guihkx @TheNexusAvenger You can temporarily fix the missing dependencies issue by installing another Flatpak which uses the same dependency (e.g. org.winehq.Wine). This is a known bug unfortunately...

I installed the new Flatpak version on the Steam Deck and it handled the extensions without any issues. The lack of a clear messaging telling me that it wouldn't be handled to the test builds + local build was annoying.
